hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/cancelairdrops/TokenCancelAirdropDecoder.java

```
@@ -0,0 +1,141 @@
+import static com.hedera.node.app.service.contract.impl.utils.ConversionUtils.asTokenId;
+import static com.hedera.node.app.spi.workflows.HandleException.validateFalse;
+import static com.hedera.node.app.spi.workflows.HandleException.validateTrue;
+
+import com.esaulpaugh.headlong.abi.Address;
+import com.esaulpaugh.headlong.abi.Tuple;
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.NftID;
+import com.hedera.hapi.node.base.PendingAirdropId;
+import com.hedera.hapi.node.base.TokenID;
+import com.hedera.hapi.node.base.TokenType;
+import com.hedera.hapi.node.token.TokenCancelAirdropTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.HtsCallAttempt;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Arrays;
+import javax.inject.Inject;
+
+public class TokenCancelAirdropDecoder {
+
+    // Tuple indexes
+    private static final int TRANSFER_LIST = 0;
+    private static final int SENDER = 0;
+    private static final int RECEIVER = 1;
+    private static final int TOKEN = 2;
+    private static final int SERIAL = 3;
+    private static final int HRC_RECEIVER = 0;
+    private static final int HRC_SERIAL = 1;
+
+    @Inject
+    public TokenCancelAirdropDecoder() {
+        // Dagger2
+    }
+
+    public TransactionBody decodeCancelAirdrop(@NonNull final HtsCallAttempt attempt) {
+        final var call = TokenCancelAirdropTranslator.CANCEL_AIRDROP.decodeCall(attempt.inputBytes());
+        final var maxPendingAirdropsToCancel =
+                attempt.configuration().getConfigData(TokensConfig.class).maxAllowedPendingAirdropsToCancel();
+        validateFalse(((Tuple[]) call.get(0)).length > maxPendingAirdropsToCancel, PENDING_AIRDROP_ID_LIST_TOO_LONG);
+
+        final var transferList = (Tuple[]) call.get(TRANSFER_LIST);
```

If L63 was above L61 you wouldn't have to do the `(cast)call.get()` twice ... (and note too that at L61 you didn't use the named constant ...)

---

Must've left it like that when I was including the fail fast mechanisms